tt_metal/api/program.hpp

```diff
@@ -0,0 +1,162 @@
+//==================================================
+
+namespace tt::tt_metal{
+namespace v1 {
+
+/**
+ * @brief Creates a Program object, which bundles kernels, circular buffers, and semaphores for execution on the device.
+ *
+ * @return Program handle to the created program.
+ */
+Program CreateProgram();
+
+
+/**
+ * @brief Creates a data movement or compute kernel and adds it to the program.
+ *
+ * @param program The program to which this kernel will be added.
+ * @param file_name Path to the kernel source file.
+ * @param core_spec Specifies the cores on which the kernel will be placed.
+ * @param config DataMovementConfig for the kernel.
+ * @return KernelHandle representing the kernel ID.
+ */
+KernelHandle CreateKernel(
+    Program program,
+    std::string_view file_name,
+    const CoreRangeSet &core_spec,
+    const DataMovementConfig &config);
+
+/**
+ * @brief Creates a data movement or compute kernel and adds it to the program.
+ *
+ * @param program The program to which this kernel will be added.
+ * @param file_name Path to the kernel source file.
+ * @param core_spec Specifies the cores on which the kernel will be placed.
+ * @param config ComputeConfig for the kernel.
+ * @return KernelHandle representing the kernel ID.
+ */
+KernelHandle CreateKernel(
+    Program program,
+    std::string_view file_name,
+    const CoreRangeSet &core_spec,
+    const ComputeConfig &config);
+
+/**
+ * @brief Creates a data movement or compute kernel and adds it to the program.
+ *
+ * @param program The program to which this kernel will be added.
+ * @param file_name Path to the kernel source file.
+ * @param core_spec Specifies the cores on which the kernel will be placed.
+ * @param config EthernetConfig for the kernel.
+ * @return KernelHandle representing the kernel ID.
+ */
+KernelHandle CreateKernel(
+    Program program,
+    std::string_view file_name,
+    const CoreRangeSet &core_spec,
+    const EthernetConfig &config);
+
+
+/**
+ * @brief Initializes a semaphore on specified cores.
+ *
+ * @param program The program to which the semaphore will be added.
+ * @param core_spec Range of cores using the semaphore.
+ * @param initial_value Initial value of the semaphore.
+ * @param core_type Core type on which to create the semaphore (default: CoreType::WORKER).
+ * @return Semaphore address as a uint32_t.
+ */
+uint32_t CreateSemaphore(
+    Program program,
+    const CoreRangeSet &core_spec,
+    uint32_t initial_value,
+    CoreType core_type = CoreType::WORKER);
+
+
+/**
+ * @brief Creates a Circular Buffer in L1 memory of specified cores and adds it to the program.
+ *
+ * @param program The program to which the buffer will be added.
+ * @param core_spec Specifies the cores where the circular buffer will be configured.
+ * @param config Configuration for the circular buffer.
+ * @return CBHandle representing the Circular Buffer ID.
+ */
+CBHandle CreateCircularBuffer(
+    Program program,
+    const CoreRangeSet &core_spec,
```

Comment:
are we deprecating creating CBs for single cores/core range

Reply:
To decrease the API surface area, we've decided not to have overloads for all three since currently the other two overloads simply construct a CoreRangeSet from the CoreCoord or CoreRange before delegating to the implementation of this overload, so having them around bought us nothing. We'll work on making CoreRangeSet cheaper and explore ways to avoid allocation when possible (using some form of SBO like std::string does).